### DISCUSSION TITLE
useRouter has unpredictable state

### DISCUSSION THREAD

**Discussion by droganov** [vercel/next.js#23737] [Help]
### What version of Next.js are you using?

latest

### What version of Node.js are you using?

14.15.0

### What browser are you using?

FF

### What operating system are you using?

MacOS

### How are you deploying your application?

devmode

### Describe the Bug

query changing it's value during component lifecycle

I have URL:
/geo?criteria=text

And the hook:
const router = useRouter()
console.log('query: ', router.query);

The bug is that query comes empty with the first pass and then populates.

The real problems is that I can not map query params to input `defaultValue`because the inputs are mounted empty and don't get the default value after the hook updates and the SSR has the wrong query value.

### Expected Behavior

query should be populated with the first pass

### To Reproduce

const router = useRouter()
console.log('query: ', router.query)

<img width="337" alt="image" src="https://user-images.githubusercontent.com/1375299/113562203-5da90a80-960e-11eb-83e1-30d67535e114.png">

**Comment by droganov** [comment]
@timneutkens why was this moved to discussion? It is a bug, the pathname is consistent and the query is not.

**Reply by droganov** [reply]
update this question is documented on https://nextjs.org/docs/api-reference/next/router#userouter

**Comment by cbdeveloper** [comment]
I'm basically asking that same question here: https://github.com/vercel/next.js/discussions/24659 and on [StackOverflow](https://stackoverflow.com/questions/67332108/next-js-router-is-returning-query-parameters-as-undefined-on-first-render)

Where did you see that on the documentation?

Do you mean this part?

![image](https://user-images.githubusercontent.com/43407798/116721218-fa27b800-a9d4-11eb-81f5-e2244d2c362c.png)

I've read that before and I was assuming it was referring to the prerender cycle (either during SSG or SSR). And I'm getting this on the client (and apparently you are also). Do you consider the render on the client to be a prerender cycle?

In my case, the page is rendered via SSG (without any data fetching for the prerender). I'm fetching data on the client, inside a `useEffect`. So I was not expecting `router.query` to be `{}` for a url like `/search?q=xxx`. I was expecting it to be `{ q: "xxx" }` since the 1st client render. I still don't get it.

**Reply by droganov** [reply]
yes, that part. After I added `getServerSideProps` query became consistent

**Reply by techrome** [reply]
@droganov Yeah when you use `getStaticProps`, it means the page will be prerendered ahead of user request so it has no way to know the query beforehand, only after the client side is initialized in the browser. The advantage of this is your server doesn't have to prerender the page for every request so it reduces the load on the server.

On the other hand, when using `getServerSideProps`, your page will be rendered on every user visit, so if, say, 1000 users go to your page, your server will have to render the page 1000 times. Of course this comes with high server load, but on the other hand the user query will be known beforehand and will be consistent.

In my opinion, it's better to get query only on client side so you don't have to make high load on the server, and while the client side is mounting, you can show some prerendered "placeholder" of the page.

But again if you really need to know the query beforehand or something depends on it, it's better to use `getServerSideProps`.

**Reply by droganov** [reply]
@techrome frankly I think that next router has a lifecycle issue with query. Technically it can see window.location at any moment on front end.

For example react-router doesn't have consistency issue.

**Reply by techrome** [reply]
@droganov It happens because your code in Next.js is first executed in Node.js environment (not the browser environment) so it doesn't have access to `window` or `document` objects and some other browser APIs. 

The code you write inside `getStaticProps` or `getServerSideProps` is executed only inside the Node.js server, not in the actual browser. Think of it like a separate environment. So technically your code is executed twice, first on the node.js server, then on the user's browser.

`react-router` doesn't have this issue because it's only executed inside browser, so it has all the access to `window` or `document` objects.

You can read more about Server Side Rendering here: [https://altalogy.com/blog/client-side-rendering-vs-server-side-rendering/](https://altalogy.com/blog/client-side-rendering-vs-server-side-rendering/)

**Reply by droganov** [reply]
@techrome 

> react-router doesn't have this issue because it's only executed inside browser, so it has all the access to window or document objects.

that's not true, react router [takes](https://reactrouter.com/web/api/StaticRouter) location from the context of StaticRouter in node and from window location in browser.

Next router could also use the context api to solve the consistency problem

**Reply by techrome** [reply]
@droganov Sorry my bad. Yeah `react-router` can read the query from user, but it's because when the user visits the page, some server-side code is being run meaning if 1000 users simultaneously open the page, the node.js code will have to be run 1000 times so the query can be fresh for every user. That's almost exactly how `getServerSideProps` work, that's why there is always fresh query because the code is run for every user.

But running `getServerSideProps` is quite expensive for the server, that's why `getStaticProps` exists - it helps you prerender your page ahead of time and then serve the generated HTML to the user, so even if million of users hit the same page only the static HTML will be served (almost). That's very cheap for the server, but the problem is there's no way to get query for the static page on the server (only when client side mounts) because the code that is run on the server is not actually run for every user. It may run only once (or in case of `revalidate` prop run only once every some seconds).

That's why `getServerSideProps` can help you get the query for the user because it will be run for every user request of the page.

**Reply by droganov** [reply]
@techrome yes, the query is unknown for the static pages during the build. So from this point the router is symmetric, the query is empty at first pass and filled SSR is enabled.

**Reply by droganov** [reply]
The case you need query is a dynamic router, because the routes fall to query. And if you need to render the links based on query, which if obviously you do, then you must do SSR, because params unknown during the build.

I think that's actually the reason behind current design. It was just strange to me because I didn't use nextjs for static pages.

**Comment by cbdeveloper** [comment]
@droganov 

Just found out a solution to this: https://github.com/vercel/next.js/discussions/24659#discussioncomment-687234

You need to check for the `router.isReady` property.

From: https://nextjs.org/docs/api-reference/next/router#router-object

![image](https://user-images.githubusercontent.com/43407798/116864659-1eb3a800-ac00-11eb-94c6-e59f1ddfd2e1.png)

I.e: If you need to use `router.query` on client for a `SSG` page. You need to "wait" for the `router.isReady` property.
